src/os/bluestore/BlueStore.h

```diff
@@ -569,10 +573,13 @@ class BlueStore : public ObjectStore,
 
   /// a cache (shard) of onodes and buffers
   struct Cache {
+    BlueStore *store;
     std::mutex lock;                ///< protect lru and other structures
 
     static Cache *create(string type);
 
+    void set_store(BlueStore *bs) { store = bs; }
```

Comment:
@ifed01 
I fully understand what you mean. I combined the above two comments you suggested into local branch but made compiler unhappy.

The code(partial):

```diff

-BlueStore::Cache *BlueStore::Cache::create(string type)
+BlueStore::Cache *BlueStore::Cache::create(string type, PerfCounters *logger)
 {
+  Cache *c;
   if (type == "lru")
-    return new LRUCache;
-  if (type == "2q")
-    return new TwoQCache;
-  assert(0 == "unrecognized cache type");
+    c = new LRUCache;
+  else if (type == "2q")
+    c = new TwoQCache;
+  else
+    assert(0 == "unrecognized cache type");
+  c->logger = logger;
+  return c;
 }

@@ -1075,10 +1087,12 @@ BlueStore::OnodeRef BlueStore::OnodeSpace::lookup(const ghobject_t& oid)
   ceph::unordered_map<ghobject_t,OnodeRef>::iterator p = onode_map.find(oid);
   if (p == onode_map.end()) {
     dout(30) << __func__ << " " << oid << " miss" << dendl;
+    logger->inc(l_bluestore_onode_misses);
     return OnodeRef();
   }
   dout(30) << __func__ << " " << oid << " hit " << p->second << dendl;
   cache->_touch_onode(p->second);
+  logger->inc(l_bluestore_onode_hits);
   return p->second;
 }

@@ -2767,7 +2814,7 @@ void BlueStore::set_cache_shards(unsigned num)
   assert(num >= old);
   cache_shards.resize(num);
   for (unsigned i = old; i < num; ++i) {
-    cache_shards[i] = Cache::create(g_conf->bluestore_cache_type);
+    cache_shards[i] = Cache::create(g_conf->bluestore_cache_type, logger);
   }
 }

@@ -569,9 +573,10 @@ public:
 
   /// a cache (shard) of onodes and buffers
   struct Cache {
+    PerfCounters *logger;
     std::mutex lock;                ///< protect lru and other structures
 
-    static Cache *create(string type);
+    static Cache *create(string type, PerfCounters *);
 
     virtual ~Cache() {}

```

It keeps complaining:
```
n file included from /home/xxg/ceph-dev/src/os/bluestore/BlueStore.cc:22:0:
/home/xxg/ceph-dev/src/os/bluestore/BlueStore.h: In member function ‘BlueStore::OnodeRef BlueStore::OnodeSpace::lookup(const ghobject_t&)’:
/home/xxg/ceph-dev/src/os/bluestore/BlueStore.h:1209:17: error: invalid use of non-static data member ‘BlueStore::logger’
   PerfCounters *logger;
                 ^
/home/xxg/ceph-dev/src/os/bluestore/BlueStore.cc:1090:5: error: from this location
     logger->inc(l_bluestore_onode_misses);
     ^
In file included from /home/xxg/ceph-dev/src/os/bluestore/BlueStore.cc:22:0:
/home/xxg/ceph-dev/src/os/bluestore/BlueStore.h:1209:17: error: invalid use of non-static data member ‘BlueStore::logger’
   PerfCounters *logger;
                 ^
/home/xxg/ceph-dev/src/os/bluestore/BlueStore.cc:1095:3: error: from this location
   logger->inc(l_bluestore_onode_hits);
   ^
```

So we can leave it this way... the difference is slight...